pkg/controller/mizar/node_controller.go

```diff
@@ -0,0 +1,307 @@
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	c := &NodeController{
+		kubeclientset:  kubeclientset,
+		informer:       informer,
+		informerSynced: informer.Informer().HasSynced,
+		lister:         informer.Lister(),
+		recorder:       recorder,
+		queue:          queue,
+	}
+	klog.Infof(c.logInfoMessage("Sending events to api server"))
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(object interface{}) {
+			resource := object.(*v1.Node)
+			key := c.genKey(resource)
+			c.Enqueue(NODE_CREATE, key)
+			fmt.Printf("NODE_CREATE: %s", key)
+		},
+		UpdateFunc: func(oldObject, newObject interface{}) {
+			oldResource := oldObject.(*v1.Node)
+			newResource := newObject.(*v1.Node)
+			oldKey := c.genKey(oldResource)
+			newKey := c.genKey(newResource)
+			if oldKey == newKey {
+				klog.Infof(c.logEventMessage("No actual change in nodes, discarding: ", newKey))
+			} else {
+				command, err := c.compareKey(oldKey, newKey)
+				if err != nil {
+					klog.Errorf(c.logInfoMessage("Unexpected string in queue; discarding: " + newKey))
+				} else {
+					c.Enqueue(command, newKey)
+				}
+				fmt.Printf("NODE_UPDATE: %s, %s", command, newKey)
+			}
+		},
+		DeleteFunc: func(object interface{}) {
+			resource := object.(*v1.Node)
+			key := c.genKey(resource)
+			c.Enqueue(NODE_DELETE, key)
+			fmt.Printf("NODE_DELETE: %s", key)
+		},
+	})
+	return c, nil
+}
+
+// Run starts an asynchronous loop that detects events of cluster nodes.
+func (c *NodeController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	klog.Infof(c.logInfoMessage("Starting node controller"))
+	klog.Infoln(c.logInfoMessage("Waiting cache to be synced"))
+	if ok := cache.WaitForCacheSync(stopCh, c.informerSynced); !ok {
+		klog.Fatalln(c.logInfoMessage("Timeout expired during waiting for caches to sync."))
+	}
+	klog.Infoln(c.logInfoMessage("Starting workers..."))
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	klog.Info(c.logInfoMessage("Shutting down node controller"))
+}
+
+// Enqueue puts key of the node object in the work queue
+func (c *NodeController) Enqueue(command string, key string) {
+	commandKey := c.genCommandKey(command, key)
+	c.queue.Add(commandKey)
+}
+
+// Dequeue an item and process it
+func (c *NodeController) runWorker() {
+	for {
+		item, queueIsEmpty := c.queue.Get()
+		if queueIsEmpty {
+			break
+		}
+		c.process(item)
+	}
+}
+
+// Parsing a item key and call gRPC request
+func (c *NodeController) process(item interface{}) {
+	defer c.queue.Done(item)
+	commandKey, ok := item.(string)
+	if !ok {
+		klog.Errorf(c.logInfoMessage("unexpected item in queue: " + commandKey))
+		c.queue.Forget(item)
```

Comment:
usually when processing the item is aborted due to error, it should put back the item into queue for later procesing. please consider to change code for robotness.